src/core/ddsi/src/ddsi_security_omg.c

```diff
@@ -686,7 +695,123 @@ static void deinit_plugin_suite_config (dds_security_plugin_suite_config *suite_
   deinit_plugin_config (&suite_config->cryptography);
 }
 
-dds_return_t q_omg_security_load (dds_security_context *sc, const dds_qos_t *qos)
+typedef bool (*expired_pp_check_fn_t)(const struct participant * pp, DDS_Security_Handle handle);
+typedef bool (*expired_proxypp_check_fn_t)(const struct proxy_participant * proxypp, DDS_Security_Handle handle);
+
+static bool delete_pp_by_handle (DDS_Security_Handle handle, expired_pp_check_fn_t expired_pp_check_fn, struct ddsi_domaingv *gv)
+{
+  struct participant *pp;
+  struct entidx_enum_participant epp;
+  bool result = false;
+  entidx_enum_participant_init (&epp, gv->entity_index);
+  while ((pp = entidx_enum_participant_next (&epp)) != NULL)
+  {
+    if (q_omg_participant_is_secure (pp) && expired_pp_check_fn (pp, handle))
+    {
+      delete_participant (gv, &pp->e.guid);
+      result = true;
+    }
+  }
+  entidx_enum_participant_fini (&epp);
+  return result;
+}
+
+static bool delete_proxypp_by_handle (const DDS_Security_Handle handle, expired_proxypp_check_fn_t expired_proxypp_check_fn, struct ddsi_domaingv *gv)
+{
+  struct proxy_participant *proxypp;
+  struct entidx_enum_proxy_participant eproxypp;
+  bool result = false;
+  entidx_enum_proxy_participant_init (&eproxypp, gv->entity_index);
+  while ((proxypp = entidx_enum_proxy_participant_next (&eproxypp)) != NULL)
+  {
+    if (q_omg_proxy_participant_is_secure(proxypp) && expired_proxypp_check_fn (proxypp, handle))
+    {
+      delete_proxy_participant_by_guid (gv, &proxypp->e.guid, ddsrt_time_wallclock (), false);
```

Comment:
The last parameter is whether the proxy sent a message explicitly stating that it was deleted by the application, or whether it was (I suppose) "implicit" because of lease expiry. I think the case of an expired certificate can be argued either way, but "implicit" and hence `true` seems slightly better to me.

Reply:
You're right, this should be an implicit delete, although the only result of setting this parameter to `true` is that the `lease_expired` flag on the proxy participant is set, which is currently not used anywhere I think. I've changed this in 23de64a